File: backend/app/schemas/curriculum_response.py
from datetime import datetime
from pydantic import RootModel, field_serializer, computed_field
from typing import List, Optional
import re

from app.config import Language, CurriculumCategory
from .base import BaseSchema
from .curriculum_create import (
    ActivitySchema,
    ExperienceSchema,
    EducationSchema,
    ProjectDescriptionSchema,
    ProjectTechnologySchema,
    ProjectSchema,
    CertificationSchema,
    StructuredCurriculumSchema,
)


class ActivityResponseSchema(ActivitySchema):
    id: str
    created_at: datetime
    updated_at: datetime

    @field_serializer("description", mode="plain")
    def serialize_bold(self, value: str) -> str:
        return re.sub(
            r"\*\*(.*?)\*\*",
            r"<strong>\1</strong>",
            value,
        )


class ExperienceResponseSchema(ExperienceSchema):
    id: str
    created_at: datetime
    updated_at: datetime

    activities: list[ActivityResponseSchema]


class EducationResponseSchema(EducationSchema):
    id: str
    created_at: datetime
    updated_at: datetime


class ProjectDescriptionResponseSchema(ProjectDescriptionSchema):
    id: str
    created_at: datetime
    updated_at: datetime

    @field_serializer("description", mode="plain")
    def serialize_bold(self, value: str) -> str:
        return re.sub(
            r"\*\*(.*?)\*\*",
            r"<strong>\1</strong>",
            value,
        )

class ProjectTechnologyResponseSchema(ProjectTechnologySchema):
    id: str
    created_at: datetime
    updated_at: datetime


class ProjectResponseSchema(ProjectSchema):
    id: str
    created_at: datetime
    updated_at: datetime

    descriptions: list[ProjectDescriptionResponseSchema]
    technologies: list[ProjectTechnologyResponseSchema] | None = None

    @computed_field
    @property
    def period(self) -> str:
        start = self.start_date.strftime("%m/%Y")

        if self.end_date is None:
            return f"{start} - Atual"

        end = self.end_date.strftime("%m/%Y")
        return f"{start} - {end}"


class CertificationResponseSchema(CertificationSchema):
    id: str
    created_at: datetime
    updated_at: datetime

    @computed_field
    @property
    def period(self) -> str:
        start = self.start_date.strftime("%m/%Y")

        if self.end_date is None:
            return f"{start} - Atual"

        end = self.end_date.strftime("%m/%Y")
        return f"{start} - {end}"

class StructuredCurriculumSummarySchema(BaseSchema):
    id: str

    language: Language
    category: CurriculumCategory

    name: str
    role: str

    email: str
    github: Optional[str] = None
    linkedin: str
    location: str

    created_at: datetime
    updated_at: datetime

class StructuredCurriculumResponseSchema(StructuredCurriculumSchema):
    id: str
    created_at: datetime
    updated_at: datetime

    experiences: list[ExperienceResponseSchema]
    educations: list[EducationResponseSchema]
    projects: list[ProjectResponseSchema] | None = None
    certifications: list[CertificationResponseSchema] | None = None

    @field_serializer("resume", mode="plain")
    def serialize_bold(self, value: str) -> str:
        return re.sub(
            r"\*\*(.*?)\*\*",
            r"<strong>\1</strong>",
            value,
        )

class ListStructuredCurriculumResponse(RootModel[List[StructuredCurriculumSummarySchema]]): ...

__all__ = [
    "ActivityResponseSchema",
    "ExperienceResponseSchema", 
    "EducationResponseSchema",
    "ProjectDescriptionResponseSchema",
    "ProjectTechnologyResponseSchema",
    "ProjectResponseSchema",
    "CertificationResponseSchema",
    "StructuredCurriculumResponseSchema",
    "StructuredCurriculumSummarySchema",
    "ListStructuredCurriculumResponse"
]

File: frontend/src/types/curriculumResponse.ts
import { CurriculumCategory, Language } from "../config/constants";

export interface CurriculumResponse {
  name: string;
  mimetype: string;
}

export interface ICurriculumResponse {
  id: string;
  language: Language;
  category: CurriculumCategory;
  name: string;
  role: string;
  email: string;
  github: string;
  linkedin: string;
  created_at: string;
  updated_at: string;
  phone: string;
  location: string;
  resume: string;
  experiences: IExperienceResponse[];
  educations: IEducationResponse[];
  projects: IProjectResponse[];
  certifications: ICertificationResponse[];
}

export interface IExperienceResponse {
  id: string;
  role: string;
  company: string;
  location: string;
  start_date: string;
  end_date: string | null;
  period: string;
  activities: IExperienceActivityResponse[];
  created_at: string;
  updated_at: string;
}

export interface IExperienceActivityResponse {
  id: string;
  description: string;
  created_at: string;
  updated_at: string;
}

export interface IEducationResponse {
  id: string;
  institution: string;
  degree: string;
  location: string;
  start_date: string;
  end_date: string | null;
  period: string;
  created_at: string;
  updated_at: string;
}

export interface IProjectResponse {
  id: string;
  name: string;
  github: string;
  demo_url: string;
  start_date: string;
  end_date: string | null;
  period: string;
  descriptions: IProjectDescriptionResponse[];
  technologies: IProjectTechnologyResponse[];
  created_at: string;
  updated_at: string;
}

export interface IProjectDescriptionResponse {
  id: string;
  description: string;
  created_at: string;
  updated_at: string;
}

export interface IProjectTechnologyResponse {
  id: string;
  technology: string;
  created_at: string;
  updated_at: string;
}

export interface ICertificationResponse {
  id: string;
  institution: string;
  name: string;
  location: string;
  start_date: string;
  end_date: string | null;
  period: string;
  created_at: string;
  updated_at: string;
}